src/planet-a/SubdivTriangleMesh.cpp

```diff
@@ -0,0 +1,170 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "SubdivTriangleMesh.h"
+
+#include <Magnum/Math/Vector3.h>
+
+using namespace planeta;
+
+
+
+ChunkedTriangleMeshInfo planeta::make_subdivtrimesh_general(
+        unsigned int chunkMax, unsigned int subdivLevels, int pow2scale)
+{
+    // calculate stuff here
+
+    using std::sqrt;
+    using std::ceil;
+    using std::pow;
+
+    float const C = chunkMax;
+    float const L = subdivLevels;
+
+    // Calculate a fair number of shared vertices, based on a triangular
+    // tiling pattern.
+    // worked out here: https://www.desmos.com/calculator/ffd8lraosl
+    unsigned int const sharedMax
+            = ceil( ( 3.0f*C + sqrt(6.0f)*sqrt(C) ) * pow(2.0f, L-1) - C + 1 );
+
+    unsigned int const fanMax
+            = ChunkedTriangleMeshInfo::smc_minFansVsLevel[subdivLevels] * 2;
+
+    return ChunkedTriangleMeshInfo(
+                chunkMax, subdivLevels, sharedMax, fanMax);
+}
+
+ChunkId ChunkedTriangleMeshInfo::chunk_create(
+        SubdivTriangleSkeleton& rSkel,
+        SkTriId skTri,
+        ArrayView_t<SkVrtxId> const edgeRte,
+        ArrayView_t<SkVrtxId> const edgeBtm,
+        ArrayView_t<SkVrtxId> const edgeLft)
+{
+    if (   edgeRte.size() != m_chunkWidth - 1
+        || edgeBtm.size() != m_chunkWidth - 1
+        || edgeLft.size() != m_chunkWidth - 1)
+    {
+        throw std::runtime_error("Incorrect edge vertex count");
+    }
+
+    // Create a new chunk ID
+    ChunkId const chunkId = m_chunkIds.create();
+    Chunk &rChunk = m_chunkData[size_t(chunkId)];
+
+    rChunk.m_skeletonTri = rSkel.tri_store(skTri);
+
+    SkeletonTriangle const &tri = rSkel.tri_at(skTri);
+
+    ArrayView_t<SharedVrtxStorage_t> const sharedSpace = chunk_shared_mutable(chunkId);
+
+    std::array const edges = {edgeRte, edgeBtm, edgeLft};
+
+    for (int i = 0; i < 3; i ++)
+    {
+        size_t const cornerOffset = m_chunkWidth * i;
+        size_t const edgeOffset = m_chunkWidth * i + 1;
+
+        ArrayView_t<SkVrtxId> const edge = edges[i];
+        {
+            SharedVrtxId const sharedId = shared_get_or_create(
+                    tri.m_vertices[i], rSkel);
+            sharedSpace[cornerOffset] = shared_store(sharedId);
+        }
+        for (unsigned int j = 0; j < m_chunkWidth - 1; j ++)
```

Comment:
prefer using fixed sized integer types. "unsigned int" can be any size, depending on the compiler in question.

Reply:
m_chunkWidth is a uint16_t. unsigned int is almost always at least 16 bits, and I'm aware that in some cases using smaller sizes can lead to worse performance. Seems like a good use for uint_fast16_t?